web/packages/teleterm/src/services/appUpdater/appUpdater.ts

```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
+    try {
+      await this.checkForUpdatesPromise;
+    } catch (error) {
+      this.logger.error('Failed to check for updates.', error);
+    } finally {
+      this.checkForUpdatesPromise = undefined;
+    }
+  }
+
+  /** Not safe for concurrent use. */
+  private async doCheckForUpdates(
+    opts: {
+      /**
+       * Whether this is a retry attempt.
+       * Used as a guard to prevent infinite loops.
+       */
+      hasRetried?: boolean;
+    } = {}
+  ): Promise<void> {
+    const result = await autoUpdater.checkForUpdates();
+
+    const newSha = result.updateInfo?.files[0]?.sha512;
+    const oldSha = this.updateCheckResult?.updateInfo.files[0]?.sha512;
+    const isSameUpdate = newSha && oldSha && newSha === oldSha;
+
+    this.updateCheckResult = result;
+
+    const updateUnavailable = !result.isUpdateAvailable;
+    const updateChanged = !isSameUpdate;
+
+    let downloadCanceled = false;
+    if (updateUnavailable || updateChanged) {
+      downloadCanceled = await this.cancelDownload();
+    }
+
+    if (
+      this.shouldAutoDownload() ||
+      // This can occur if the user manually downloads an update
+      // and then triggers another check for updates.
+      // Since the update is already downloaded, the updater should transition
+      // to the `update-downloaded` state automatically.
+      // The update file will be read from the local cache.
+      this.isUpdateDownloaded
+    ) {
+      void this.download();
+      return;
+    }
+
+    // Retry to refresh the state so that the UI won't be showing
+    // a cancellation error.
+    if (downloadCanceled && !opts.hasRetried) {
+      await this.doCheckForUpdates({ hasRetried: true });
+    }
+  }
+
+  /** Starts download. */
+  async download(): Promise<string[]> {
+    if (this.downloadPromise) {
+      this.logger.info('Download already in progress.');
+      return this.downloadPromise;
+    }
+
+    this.downloadPromise = autoUpdater.downloadUpdate();
+    try {
+      await this.downloadPromise;
+      this.isUpdateDownloaded = true;
+    } catch (error) {
+      this.logger.error('Failed to download update.', error);
+    } finally {
+      this.downloadPromise = undefined;
+    }
+  }
+
+  /** Cancels download. Returns true if aborted the network request. */
+  async cancelDownload(): Promise<boolean> {
+    if (!this.downloadPromise) {
+      this.isUpdateDownloaded = false;
+      return false;
+    }
+
+    // Due to a bug in electron-updater, we can't cancel downloads using cancellation
+    // token passed to autoUpdater.download().
+    // Repeatedly starting and canceling downloads causes the updater to go
+    // into a broken state where it becomes unresponsive.
+    // To avoid this, we instead close the network connections to abort
+    // the current download.
+    try {
+      await autoUpdater.netSession.closeAllConnections();
+      await this.downloadPromise;
+      return false;
+    } catch {
+      return true;
```

Comment:
Is there a way for this `catch` to be more specific rather than returning `true` for all errors? Or maybe at least wrap just `this.downloadPromise` in `try catch`?

I assume that we don't care about the error from `this.downloadPromise` and just want to wait until the promise fails – a very broad `try catch` like that is understandable in this case. But we expect `closeAllConnections` to not throw, right? If it fails, it feels like the whole method should fail too.